### PR TITLE
RD-1766 refactor pipeline, parallel tests and add caching

### PR DESCRIPTION
**Pull request by yuvlar**
The changes don't follow the exact diagram presented by @qooban in RD-1766 for two reasons:
1. The Loki tests require the storybook to run first.
2. Running all tests in parallel chokes the node and slows the build.

I also changed the node selector to use spot-xlarge as it reduces the runtime in almost half (5-6min avg vs 8-10min), and set a low CPU request so other pods can utilize the node as well.

### CODE REVIEW

File: jenkins/Jenkinsfile [hunk 1]
```diff
@@ -36,41 +45,79 @@ pipeline {
     timestamps()
   }
   stages {
-    stage('Build') {
-      steps {
-        container('node'){
-          dir("${env.WORKSPACE}/${env.PROJECT}") {
-            echo "Install dependencies"
-            sh 'npm ci --prefer-offline'
+    stage('CI & Audit'){
```
Comment (qooban) [inline jenkins/Jenkinsfile:48]: Maybe
```suggestion
    stage('Install & Audit'){
```
would be more accurate? (as CI might be misleading)

File: jenkins/Jenkinsfile [hunk 2]
```diff
@@ -36,41 +45,79 @@ pipeline {
...
+                echo "Install dependencies"
+                sh 'npm ci --prefer-offline'
+                archiveFolder('~/.npm', "${env.WORKSPACE}/${env.PROJECT}/package-lock.json", 'components-mb', 'npm_dir', '/mnt/data')
+              }
+            }
+          }
+        }
+      }
+    }
+    stage('Run Build & Tests'){
```
Comment (qooban) [inline jenkins/Jenkinsfile:77]: Maybe to keep the same naming convention for all stages we shouldn't use "Run"?
```suggestion
    stage('Build & Tests'){
```

File: jenkins/Jenkinsfile [hunk 3]
```diff
@@ -36,41 +45,79 @@ pipeline {
...
+                sh 'npm run build'
+
+                echo "Build Storybook"
+                sh 'npm run build-storybook'
+              }
+            }
+          }
+        }
+
+        stage('Run lint & check-types & jest'){
```
Comment (qooban) [inline jenkins/Jenkinsfile:93]: I would name it:
```suggestion
        stage('Static analysis & Unit tests'){
```

File: jenkins/Jenkinsfile [hunk 4]
```diff
@@ -36,41 +45,79 @@ pipeline {
...
+                echo "Run Jest tests"
+                sh 'npm run test:jest'
+              }
+            }
           }
         }
       }
     }
-    stage('Test') {
+    stage('Loki Test') {
```
Comment (qooban) [inline jenkins/Jenkinsfile:111]: Maybe
```suggestion
    stage('Visual tests') {
```
?

**Review by qooban — COMMENTED** [review]
Added some comments.

One more general (just not to forget) -  we'll also need to update GitHub stages (required / optional).

File: jenkins/Jenkinsfile [hunk 5]
```diff
@@ -1,3 +1,10 @@
+def getMainBranch(String branch){
+  if (branch =~ /.+-build/)
+    return branch
+  else
+    return 'master'
+}
```
[the remark this answers is not in the document]
Comment (yuvlar) [reply jenkins/Jenkinsfile:6]: copy that, I'll remove it

File: jenkins/Jenkinsfile [hunk 6]
```diff
@@ -36,41 +45,79 @@ pipeline {
...
+          }
+          steps {
+            container('node'){
+              dir("${env.WORKSPACE}/${env.PROJECT}") {
+                echo 'Run npm production dependencies security audit'
+                sh 'npm run audit'
+              }
+            }
+          }
+        }
```
[the remark this answers is not in the document]
Comment (yuvlar) [reply jenkins/Jenkinsfile:62]: Sounds good, I'll give it a try

File: jenkins/Jenkinsfile [hunk 7]
```diff
@@ -36,41 +45,79 @@ pipeline {
...
-            echo "Check TypeScript types"
-            sh "npm run check-types"
-
             echo "Run Loki tests"
             sh '''
               apt-get update && apt-get -y install graphicsmagick
               google-chrome --version
             '''
-            catchError(buildResult: 'FAILURE', stageResult: 'FAILURE') {
+            catchError {
```
[the remark this answers is not in the document]
Comment (yuvlar) [reply jenkins/Jenkinsfile:120]: Hmm, didn't notice I removed it, thanks.

File: jenkins/Jenkinsfile (hunk 1 above)
Comment (yuvlar) [reply jenkins/Jenkinsfile:48]: I'll take your suggestion

File: jenkins/Jenkinsfile (hunk 4 above)
Comment (qooban) [reply jenkins/Jenkinsfile]: @yuvlar I see you updated it, but let's have here `Visual tests` instead of `Visual Test`.

**Review by qooban — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**